homework/grayscale-image/compression.cpp

```diff
@@ -0,0 +1,77 @@
+#include "compression.hpp"
+#include <algorithm>
+#include <iostream>
+
+// old version without STL
+// CompressedBitmap compressGrayscale(const Bitmap& bitmap) {
+//     CompressedBitmap compressed{};
+//     compressed.reserve(width * height);
+
+//     uint8_t currCount{1};
+//     for (Bitmap::size_type i{0}; i < height; ++i) {
+//         for (Bitmap::size_type j{1}; j < width + 1; ++j) {
+//             while (j < width && bitmap[i][j] == bitmap[i][j - 1]) {
+//                 ++currCount;
+//                 ++j;
+//             }
+//             compressed.emplace_back(bitmap[i][j - 1], currCount);
+//             currCount = 1;
+//         }
+//     }
+
+//     compressed.shrink_to_fit();
+//     return compressed;
+// }
+
+CompressedBitmap compressGrayscale(const Bitmap& bitmap) {
+    CompressedBitmap compressed{};
+    compressed.reserve(height * width);
+    for (auto row : bitmap) {
+        for (auto currentElement = begin(row), newElement = begin(row); currentElement != end(row); ) {
+            newElement = std::find_if(currentElement,
+                                      end(row),
+                                      [&](auto value) {
+                                          return value != *currentElement;
+                                      });
+            compressed.emplace_back(*currentElement, std::distance(currentElement, newElement));
+            currentElement = newElement;
+        }
+    }
+    compressed.shrink_to_fit();
+    return compressed;
+}
+
+Bitmap decompressGrayscale(const CompressedBitmap& compressed) {
+    Bitmap bitmap{};
+
+    auto it = bitmap.begin();
+    int count{0};
+    for (const auto& [Val, N] : compressed) {
+        std::fill_n(it->begin() + count, N, Val);
+        count += N;
+        if (count >= width) {
+            ++it;
+            count = 0;
+        }
```

Comment:
Use iterator returned from std::fill_n. Vector is conigiuous so all elemnts are in the same palce. When you handle one row, the next element is the begining of the second row. So you just need to keep track an iterator and pass it,

`it = std::fill_n(it, N, val);`

Vector2D: {{x, x, x}, {y, y,y}, {z,z,z}
Is sotred in emory like this:
0  1  2  3  4  5  6  7  8
[x][x][x][y][y][y][z][z][z]